File: csharp/config/store/Domain.cs
/* 
 Copyright (c) 2010, Direct Project
 All rights reserved.

 Authors:
    Umesh Madan     umeshma@microsoft.com
  
Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
Neither the name of The Direct Project (directproject.org) nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
*/
using System;
using System.Data.Linq.Mapping;
using System.Runtime.Serialization;
using System.Net.Mail;

namespace Health.Direct.Config.Store
{
    [Table(Name="Domains")]
    [DataContract(Namespace = ConfigStore.Namespace)]
    public class Domain
    {
        public const int MaxDomainNameLength = 255;
        public const int MaxGroupNameLength = 25;

        string m_name;
        string m_agentName;
        
        public Domain()
        {
            this.CreateDate = DateTimeHelper.Now;
            this.UpdateDate = this.CreateDate;
            this.Status = EntityStatus.New;
        }
        
        public Domain(string name) : this()
        {
            this.Name = name;
        }
           
        [Column(Name="DomainID", IsPrimaryKey=true, IsDbGenerated=true, UpdateCheck = UpdateCheck.Never)]
        [DataMember(IsRequired = true)]
        public long ID
        {
            get;
            set;
        }

        [Column(Name = "DomainName", DbType = "varchar(255)", CanBeNull = false, IsPrimaryKey = true, UpdateCheck = UpdateCheck.Never)]
        [DataMember(IsRequired = true)]
        public string Name
        {
            get
            {
                return m_name;
            }
            set
            {
                if (string.IsNullOrEmpty(value))
                {
                    throw new ConfigStoreException(ConfigStoreError.InvalidDomainName);
                }
                
                if (value.Length > MaxDomainNameLength)
                {
                    throw new ConfigStoreException(ConfigStoreError.DomainNameLength);
                }
                                
                m_name = value;
            }
        }

        [Column(Name = "AgentName", DbType = "varchar(25)", CanBeNull = true, IsPrimaryKey = false, UpdateCheck = UpdateCheck.Never)]
        [DataMember(IsRequired = false)]
        public string AgentName
        {
            get
            {
                return m_agentName;
            }
            set
            {
                value = value ?? string.Empty;
                if (value.Length > MaxGroupNameLength)
                {
                    throw new ConfigStoreException(ConfigStoreError.AgentNameLength);
                }

                m_agentName = value;
            }
        }
        
        [Column(Name = "CreateDate", CanBeNull = false, UpdateCheck = UpdateCheck.Never)]
        [DataMember(IsRequired = true)]
        public DateTime CreateDate
        {
            get;
            set;
        }

        [Column(Name = "UpdateDate", CanBeNull = false, UpdateCheck = UpdateCheck.Always)]
        [DataMember(IsRequired = true)]
        public DateTime UpdateDate
        {
            get;
            set;
        }

        [Column(Name = "Status", DbType = "tinyint", CanBeNull = false, UpdateCheck = UpdateCheck.Never)]
        [DataMember(IsRequired = true)]
        public EntityStatus Status
        {
            get;
            set;
        }

        internal void CopyFixed(Domain source)
        {
            this.ID = source.ID;
            this.CreateDate = source.CreateDate;
            this.Name = source.Name;
            this.UpdateDate = source.UpdateDate;
        }
        
        internal void ApplyChanges(Domain source)
        {
            this.Status = source.Status;
            this.AgentName = source.AgentName;
            this.UpdateDate = DateTimeHelper.Now;
        }
        
        public bool IsValidEmailDomain()
        {
            return IsValidEmailDomain(this.Name);
        }        
        /// <summary>
        /// The robust way to validate that a domainName meets the criteria of RFC5322...is to parse it as as an Address Field.
        /// We use MailAdress to do that.
        /// </summary>
        public static bool IsValidEmailDomain(string domainName)
        {
            if (string.IsNullOrEmpty(domainName))
            {
                throw new ArgumentException("value was null or empty", "domainName");
            }
            try
            {
                MailAddress address = new MailAddress("unknown.user@" + domainName);
                return address.Host.Equals(domainName, StringComparison.OrdinalIgnoreCase);
            }
            catch
            {
            }
            
            return false;
            
        }
    }
}

File: java/config/config-model/src/main/java/org/nhindirect/config/model/Domain.java
/* 
Copyright (c) 2010, NHIN Direct Project
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer 
   in the documentation and/or other materials provided with the distribution.  
3. Neither the name of the The NHIN Direct Project (nhindirect.org) nor the names of its contributors may be used to endorse or promote 
   products derived from this software without specific prior written permission.
   
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, 
THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS 
BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE 
GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, 
STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
THE POSSIBILITY OF SUCH DAMAGE.
 */

package org.nhindirect.config.model;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collection;
import java.util.Collections;

import org.codehaus.enunciate.json.JsonRootType;

/**
 * A Direct email domain.
 * @author Greg Meyer
 * @since 1.0
 */
///CLOVER:OFF
@JsonRootType
public class Domain 
{
    private String domainName;

    private Calendar createTime;

    private Calendar updateTime;

    private Address postmasterAddress;

    private Collection<Address> addresses;

    private long id;

    private EntityStatus status;
    
    /**
     * Empty constructor
     */
    public Domain()
    {
    	
    }

    /**
     * Gets the name of the domain.
     * @return The name of the domain.
     */
	public String getDomainName() 
	{
		return domainName;
	}

	/**
	 * Sets the name of the domain.
	 * @param domainName The name of the domain.
	 */
	public void setDomainName(String domainName) 
	{
		this.domainName = domainName;
	}

	/**
	 * Sets the date/time the domain was created in the system.
	 * @return The date/time the domain was created in the system.
	 */
	public Calendar getCreateTime() 
	{
		return createTime;
	}

	/**
	 * Sets the date/time the domain was created in the system.
	 * @param createTime The date/time the domain was created in the system.
	 */ 
	public void setCreateTime(Calendar createTime) 
	{
		this.createTime = createTime;
	}

	/**
	 * Gets the date/time that the domain was last updated.
	 * @return The date/time that the domain was last updated.
	 */
	public Calendar getUpdateTime() 
	{
		return updateTime;
	}

	/**
	 * Sets the date/time that the domain was last updated.
	 * @param updateTime The date/time that the domain was last updated.
	 */
	public void setUpdateTime(Calendar updateTime) 
	{
		this.updateTime = updateTime;
	}

	/**
	 * Gets the email address of the domain's postmaster.
	 * @return The email address of the domain's postmaster.
	 */
	public Address getPostmasterAddress() 
	{
		return postmasterAddress;
	}

	/**
	 * Sets the email address of the domain's postmaster.
	 * @param postmasterAddress The email address of the domain's postmaster.
	 */
	public void setPostmasterAddress(Address postmasterAddress) 
	{
		this.postmasterAddress = postmasterAddress;
	}

	/**
	 * Gets all addresses that are configured for a domain.
	 * @return All addresses that are configured for a domain.
	 */
	public Collection<Address> getAddresses() 
	{
		if (addresses == null)
			addresses = Collections.emptyList();
		
		return Collections.unmodifiableCollection(addresses);
	}

	/**
	 * Sets all addresses that are configured for a domain.
	 * @param addresses All addresses that are configured for a domain.
	 */
	public void setAddresses(Collection<Address> addresses) 
	{
		this.addresses = new ArrayList<Address>(addresses);
	}

	/**
	 * Gets the internal system id of the domain.
	 * @return The internal system id of the domain.
	 */
	public long getId() 
	{
		return id;
	}

	/**
	 * Sets the internal system id of the domain.
	 * @param id The internal system id of the domain.
	 */
	public void setId(long id) 
	{
		this.id = id;
	}

	/**
	 * Gets the status of the domain.
	 * @return The status of the domain.
	 */
	public EntityStatus getStatus() 
	{
		return status;
	}

	/**
	 * Sets the status of the domain.
	 * @param status The status of the domain.
	 */
	public void setStatus(EntityStatus status) 
	{
		this.status = status;
	}
    
    
}
///CLOVER:ON
